src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs

```diff
@@ -0,0 +1,173 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Blockchain.Find;
+using Nethermind.Core;
+using Nethermind.Int256;
+
+namespace Nethermind.JsonRpc.Modules.Eth
+{
+    public class GasPriceOracle : IGasPriceOracle
+    {
+        public UInt256? DefaultGasPrice { get; private set; }
+        public List<UInt256> TxGasPriceList { get; private set; }
+        private Block? _lastHeadBlock;
+        private UInt256? _lastGasPrice;
+        private IBlockFinder? _blockFinder;
+        private EarlyExitManager? _earlyExitManager;
+        private bool _eip1559Enabled;
+        private readonly UInt256? _ignoreUnder;
+        private readonly int _blockLimit;
+        private readonly int _softTxThreshold;
+        private readonly UInt256 _baseFee;
+        private readonly ValidTxInsertionManager _addValidTx;
+
+        public GasPriceOracle(bool eip1559Enabled = false, UInt256? ignoreUnder = null, 
+            int? blockLimit = null, UInt256? baseFee = null)
+        {
+            TxGasPriceList = new List<UInt256>();
+            _eip1559Enabled = eip1559Enabled;
+            _ignoreUnder = ignoreUnder ?? UInt256.Zero;
+            _blockLimit = blockLimit ?? GasPriceConfig.DefaultBlocksLimit;
+            _softTxThreshold = GasPriceConfig.SoftTxLimit;
+            _baseFee = baseFee ?? GasPriceConfig.DefaultBaseFee;
+            _addValidTx = new ValidTxInsertionManager(this, _ignoreUnder, _eip1559Enabled, _baseFee);
+            _earlyExitManager = null;
+        }
+
+        public ResultWrapper<UInt256?> GasPriceEstimate(IBlockFinder blockFinder)
+        {
+            if (_blockFinder == null)
```

Comment:
why do you pass this argument in the estimate method and then set it as a field?
either set it in the constructor or pass around?
otherwise this is very not thread safe?